nixos/matrix-sliding-sync: add dependency on matrix-synapse if running locally

999eagle opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

This makes sure `matrix-sliding-sync` has a dependency on `matrix-synapse` if both are configured on the same host and configures `matrix-sliding-sync` to automatically restart as the service fails if synapse is not reachable yet.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

999eagle committed:
nixos/matrix-sliding-sync: add dependency on matrix-synapse if running locally and restart

Ma27 commented on nixos/modules/services/matrix/matrix-sliding-sync.nix:
```diff
@@ -80,8 +80,11 @@ in
+      after =
+        lib.optional cfg.createDatabase "postgresql.service"
+        ++ lib.optional config.services.matrix-synapse.enable "matrix-synapse.service";
+      wants = after;
```
Don't you need `requires` to propagate a restart (that's how I understand your PR describe at least):

> [...] configures matrix-sliding-sync to automatically restart as the service fails if synapse is not reachable yet.

Also, did the service actually break? I cannot reproduce this, here are the logs from `matrix-sliding-sync.service` when doing a `systemctl restart matrix-synapse.target`:

```
Nov 04 13:44:50 soost syncv3[2172]: 13:44:50 WRN Poller: sync v2 poll returned temporary error error="DoSyncV2: response returned 502 Bad Gateway" code=502 device_id=YDVFFUECYG user_id=@ma27:nicht-so.sexy
Nov 04 13:44:50 soost syncv3[2172]: 13:44:50 WRN Poller: waiting before next poll device_id=YDVFFUECYG duration=3s fail-count=1 user_id=@ma27:nicht-so.sexy
```

In other words I'm not sure if this would be desirable.

999eagle replied:
The automatic restart is meant to describe the `Restart` and `RestartSec` options set in the service.
I don't have logs right now but I did have sporadic issues with `matrix-sliding-sync.service` failing to start because it couldn't reach the upstream homeserver. On deployments where both the sync proxy and the homeserver are running on the same host, `Wants` and `After` make sure the sync proxy isn't started without the homeserver (though we could change this to `Requires` if we want to). On deployments where sync proxy and homeserver are running on different hosts, the `Restart` and `RestartSec` make sure the sync proxy is restarted automatically when it fails.

Ma27 replied:
> The automatic restart is meant to describe the Restart and RestartSec options set in the service

OK never mind then, I misunderstood you then, I thought this was about actions to the synapse service being propagated to sliding sync. Fwiw, I agree about after/wants here.

> but I did have sporadic issues with matrix-sliding-sync.service failing to start because it couldn't reach the upstream homeserver

The example above is about a restart only, I haven't checked closely how it behaves on startup.

Nothing else to discuss here IMHO, so I'll resolve.

Ma27 commented on nixos/modules/services/matrix/matrix-sliding-sync.nix:
```diff
@@ -80,8 +80,11 @@ in
+    systemd.services.matrix-sliding-sync = rec {
+      after =
+        lib.optional cfg.createDatabase "postgresql.service"
+        ++ lib.optional config.services.matrix-synapse.enable "matrix-synapse.service";
```
Does it fail because synapse itself is unreachable or because of $worker not being there? If the latter is the case, we'd probably want to use `matrix-synapse.target` here.

999eagle replied:
I'm not actually sure which exact service is required. I've chosen to depend on `matrix-synapse.service` because that's always present unlike `matrix-synapse.target` which only exists when using workers and this way the configuration becomes a bit simpler. This is also how other services declare their dependency on Synapse (see [mautrix-whatsapp](https://github.com/NixOS/nixpkgs/blob/fa804edfb7869c9fb230e174182a8a1a7e512c40/nixos/modules/services/matrix/mautrix-whatsapp.nix#L103) or [appservice-discord](https://github.com/NixOS/nixpkgs/blob/fa804edfb7869c9fb230e174182a8a1a7e512c40/nixos/modules/services/matrix/appservice-discord.nix#L103) for examples).

Ma27 replied:
>  (see mautrix-whatsapp or appservice-discord for examples)

One of the reasons is probably that these modules are older than worker support.

> which only exists when using workers and this way the configuration becomes a bit simpler

What about checking if we have workers on the system and use either target or service?

Unless the sliding sync proxy strictly depends on the main process only no matter how/what workers are configured of course.

999eagle replied:
I've added a new option `config.services.matrix-synapse.serviceUnit` now which is either `matrix-synapse.service` or `matrix-synapse.target`. That way, other modules don't need to check for worker support and can just add an `After`/`Wants` dependency on the new option. Do you think this is a good idea?

Ma27 replied:
Yes, I like that :+1:

Ma27 commented on nixos/modules/services/matrix/matrix-sliding-sync.nix:
```diff
@@ -90,6 +93,8 @@ in
         StateDirectory = "matrix-sliding-sync";
         WorkingDirectory = "%S/matrix-sliding-sync";
+        Restart = "on-failure";
+        RestartSec = "5s";
```
With the default burst (5) and the default start limit interval (10s) this will never reach a failing state (see `systemd-system.conf(5)` for the global defaults and `systemd.unit(5)` for an explanantion), but will retry infinitely. I don't think that this is desirable since this effectively means that you need dedicated monitoring for this service (otherwise, checking if a unit reached the failing state was sufficient, but this won't be happening here AFAIU).

I don't have a strong opinion on whether to increase the StartLimitTimeoutSec or to decrease the restart sec here, but I'd prefer to have it implemented that way rather than having infinite retries. If somebody has an outstandingly slow Matrix setup which would regularly run into that timeout, I'd expect them to update the service's configuration anyways.

999eagle replied:
Ah yes, that's right. I missed that. I guess we could just decrease `RestartSec` to `1s` or remove it outright to be able to reach a failing state. Do you have any preference on which to use?

Ma27 replied:
I'd say that 1s is OK (when starting up the proxy has to wait for synapse to be up anyways because of Wants= and I haven't even seen a single crash when restarting synapse itself).

This is a good step forward and an improvement over the previous state. If it turns out that this was too optimistic we can fix it later as well.

999eagle replied:
Changed to 1s

999eagle committed:
nixos/matrix-synapse: add readOnly serviceUnit option

This is useful to configure services dependent on matrix-synapse regardless
of whether workers are enabled or not.

999eagle committed:
nixos/matrix/*: change dependencies on matrix-synapse.service to serviceUnit

Ma27 commented:
Thanks!
